# Wrangling and Analysis of WeRateDogs Twitter Archive by Cady Patino

## Gather

In [73]:
# load packages
import pandas as pd
import tweepy
import json
import numpy as np
import requests
import os

I first want to read in the twitter archive csv file provided by Udacity.

In [74]:
# read twitter archive csv file into dataframe
twitter_archive = pd.read_csv('twitter-archive-enhanced.csv')

Next I want to programatically download the image predictions file from Udacity's servers.

In [75]:
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)

# confirm the request has succeeded (check http status code)
response

<Response [200]>

In [76]:
# create empty tsv vile and write contents of response to it
with open('image_predictions.tsv', mode ='wb') as file:
    file.write(response.content)

In [77]:
# read tsv file into dataframe
image_predictions = pd.read_csv('image_predictions.tsv', sep='\t')

Lastly, I want to gather additional data about these tweets using Twitter's API.

In [78]:
import tweepy

consumer_key = 'nIwdyYG87k7KOZ4dtjvHoBksf'
consumer_secret = 'jqXscVh0wqZKLWCse9TNK4YY3t8KTXQslRFN0kmmINbBASTSta'
access_token = '80207001-EWsvq4Rmvrfs9zmlGP5r8CWcGM5CTdf9VfTMvj7Mi'
access_secret = 'gnVNzdL424JMNl5qsHBA0BAexcqze3BSJM5xITaiB0LwT'

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

# set authentication handler and wait parameters
api = tweepy.API(auth, wait_on_rate_limit = True, wait_on_rate_limit_notify = True)

In [81]:
# get list of all tweet ids
tweet_ids = twitter_archive['tweet_id'].values.tolist()

id_errors = [] # initialize list for ids with errors

# get tweet info (json) for all tweet ids in list and keep track of errors
for tweet_id in tweet_ids:
    try:
        tweet = api.get_status(tweet_id, tweet_mode = 'extended')
        print(tweet_id)
        with open('tweet_json.txt', 'a') as file:
            json.dump(tweet._json, file)
            file.write('\n')
    except tweepy.TweepError as e:
        print (e.reason)
        id_errors.append(tweet_id)

892420643555336193
892177421306343426
891815181378084864
891689557279858688
891327558926688256
891087950875897856
890971913173991426
890729181411237888
890609185150312448
890240255349198849
890006608113172480
889880896479866881
889665388333682689
889638837579907072
889531135344209921
889278841981685760
888917238123831296
888804989199671297
888554962724278272
[{'code': 144, 'message': 'No status found with that ID.'}]
888078434458587136
887705289381826560
887517139158093824
887473957103951883
887343217045368832
887101392804085760
886983233522544640
886736880519319552
886680336477933568
886366144734445568
886267009285017600
886258384151887873
886054160059072513
885984800019947520
885528943205470208
885518971528720385
885311592912609280
885167619883638784
884925521741709313
884876753390489601
884562892145688576
884441805382717440
884247878851493888
884162670584377345
883838122936631299
883482846933004288
883360690899218434
883117836046086144
882992080364220416
882762694511734784
882627270

Rate limit reached. Sleeping for: 232


878057613040115712
877736472329191424
877611172832227328
877556246731214848
877316821321428993
877201837425926144
876838120628539392
876537666061221889
876484053909872640
876120275196170240
875747767867523072
875144289856114688
875097192612077568
875021211251597312
874680097055178752
874434818259525634
874296783580663808
874057562936811520
874012996292530176
[{'code': 144, 'message': 'No status found with that ID.'}]
873580283840344065
873337748698140672
873213775632977920
872967104147763200
872820683541237760
872668790621863937
872620804844003328
872486979161796608
872261713294495745
872122724285648897
871879754684805121
871762521631449091
871515927908634625
871166179821445120
871102520638267392
871032628920680449
870804317367881728
870726314365509632
870656317836468226
870374049280663552
870308999962521604
870063196459192321
[{'code': 144, 'message': 'No status found with that ID.'}]
869772420881756160
869702957897576449
869596645499047938
869227993411051520
868880397819494401
868639

813812741911748608
813800681631023104
813217897535406080
813202720496779264
813187593374461952
813172488309972993
813157409116065792
813142292504645637
813130366689148928
813127251579564032
813112105746448384
813096984823349248
813081950185472002
813066809284972545
813051746834595840
812781120811126785
812747805718642688
812709060537683968
812503143955202048
812466873996607488
812372279581671427
811985624773361665
811744202451197953
811647686436880384
811627233043480576
811386762094317568
810984652412424192
810896069567610880
810657578271330305
810284430598270976
810254108431155201
809920764300447744
809808892968534016
809448704142938112
809220051211603969
809084759137812480
808838249661788160
808733504066486276
808501579447930884
808344865868283904
808134635716833280
808106460588765185
808001312164028416
807621403335917568
807106840509214720
807059379405148160
807010152071229440
806629075125202948
806620845233815552
806576416489959424
806542213899489280
806242860592926720
806219024703

756939218950160384
756651752796094464
756526248105566208
756303284449767430
756288534030475264
756275833623502848
755955933503782912
755206590534418437
755110668769038337
754874841593970688
754856583969079297
754747087846248448
754482103782404096
754449512966619136
754120377874386944
754011816964026368
753655901052166144
753420520834629632
753398408988139520
753375668877008896
753298634498793472
753294487569522689
753039830821511168
753026973505581056
752932432744185856
752917284578922496
752701944171524096
752682090207055872
752660715232722944
752568224206688256
752519690950500352
752334515931054080
752309394570878976
752173152931807232
751950017322246144
751937170840121344
751830394383790080
751793661361422336
751598357617971201
751583847268179968
751538714308972544
751456908746354688
751251247299190784
751205363882532864
751132876104687617
750868782890057730
750719632563142656
750506206503038976
750429297815552001
750383411068534784
750381685133418496
750147208377409536
750132105863

Rate limit reached. Sleeping for: 619


750026558547456000
750011400160841729
749996283729883136
749981277374128128
749774190421639168
749417653287129088
749403093750648834
749395845976588288
749317047558017024
749075273010798592
749064354620928000
749036806121881602
748977405889503236
748932637671223296
748705597323898880
748699167502000129
748692773788876800
748575535303884801
748568946752774144
748346686624440324
748337862848962560
748324050481647620
748307329658011649
748220828303695873
747963614829678593
747933425676525569
747885874273214464
747844099428986880
747816857231626240
747651430853525504
747648653817413632
747600769478692864
747594051852075008
747512671126323200
747461612269887489
747439450712596480
747242308580548608
747219827526344708
747204161125646336
747103485104099331
746906459439529985
746872823977771008
746818907684614144
746790600704425984
746757706116112384
746726898085036033
746542875601690625
746521445350707200
746507379341139972
746369468511756288
746131877086527488
746056683365994496
745789745784

699072405256409088
699060279947165696
699036661657767936
698989035503689728
698953797952008193
698907974262222848
698710712454139905
698703483621523456
698635131305795584
698549713696649216
698355670425473025
698342080612007937
698262614669991936
698195409219559425
698178924120031232
697995514407682048
697990423684476929
697943111201378304
697881462549430272
697630435728322560
697616773278015490
697596423848730625
697575480820686848
697516214579523584
697482927769255936
697463031882764288
697270446429966336
697259378236399616
697255105972801536
697242256848379904
696900204696625153
696894894812565505
696886256886657024
696877980375769088
696754882863349760
696744641916489729
696713835009417216
696518437233913856
696490539101908992
696488710901260288
696405997980676096
696100768806522880
695816827381944320
695794761660297217
695767669421768709
695629776980148225
695446424020918272
695409464418041856
695314793360662529
695095422348574720
695074328191332352
695064344191721472
695051054296

675888385639251968
675878199931371520
675870721063669760
675853064436391936
675849018447167488
675845657354215424
675822767435051008
675820929667219457
675798442703122432
675781562965868544
675740360753160193
675710890956750848
675707330206547968
675706639471788032
675534494439489536
675531475945709568
675522403582218240
675517828909424640
675501075957489664
675497103322386432
675489971617296384
675483430902214656
675432746517426176
675372240448454658
675362609739206656
675354435921575936
675349384339542016
675334060156301312
675166823650848770
675153376133427200
675149409102012420
675147105808306176
675146535592706048
675145476954566656
675135153782571009
675113801096802304


Rate limit reached. Sleeping for: 608


675111688094527488
675109292475830276
675047298674663426
675015141583413248
675006312288268288
675003128568291329
674999807681908736
674805413498527744
674800520222154752
674793399141146624
674790488185167872
674788554665512960
674781762103414784
674774481756377088
674767892831932416
674764817387900928
674754018082705410
674752233200820224
674743008475090944
674742531037511680
674739953134403584
674737130913071104
674690135443775488
674670581682434048
674664755118911488
674646392044941312
674644256330530816
674638615994089473
674632714662858753
674606911342424069
674468880899788800
674447403907457024
674436901579923456
674422304705744896
674416750885273600
674410619106390016
674394782723014656
674372068062928900
674330906434379776
674318007229923329
674307341513269249
674291837063053312
674271431610523648
674269164442398721
674265582246694913
674262580978937856
674255168825880576
674082852460433408
674075285688614912
674063288070742018
674053186244734976
674051556661161984
674045139690

666817836334096384
666804364988780544
666786068205871104
666781792255496192
666776908487630848
666739327293083650
666701168228331520
666691418707132416
666649482315059201
666644823164719104
666454714377183233
666447344410484738
666437273139982337
666435652385423360
666430724426358785
666428276349472768
666421158376562688
666418789513326592
666411507551481857
666407126856765440
666396247373291520
666373753744588802
666362758909284353
666353288456101888
666345417576210432
666337882303524864
666293911632134144
666287406224695296
666273097616637952
666268910803644416
666104133288665088
666102155909144576
666099513787052032
666094000022159362
666082916733198337
666073100786774016
666071193221509120
666063827256086533
666058600524156928
666057090499244032
666055525042405380
666051853826850816
666050758794694657
666049248165822465
666044226329800704
666033412701032449
666029285002620928
666020888022790149


In [84]:
# read json into dataframe
additional_twitter_info = pd.read_json('tweet_json.txt', lines=True)

In [85]:
# save dataframe to csv file
additional_twitter_info.to_csv('additional_twitter_info.csv')

In [86]:
# count number of tweet ids that no longer exist
len(id_errors)

11

In [87]:
# show ids that no longer exist
id_errors

[888202515573088257,
 873697596434513921,
 869988702071779329,
 866816280283807744,
 861769973181624320,
 845459076796616705,
 842892208864923648,
 837012587749474308,
 827228250799742977,
 802247111496568832,
 775096608509886464]

## Assess

First, I'll assess the `twitter_archive` dataframe and I'll record the quality and tidiness issues I find.

In [88]:
twitter_archive

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

From my initial visual assessment, I can see that some names are wrong or listed as 'None' (quality issue). Also, the timestamp column has day, month, and year all in the same column instead of parsed (tidiness issue). Lastly, I can see that some of the entries are retweets and should not be used in the analysis - the text for these entries begins with 'RT'. These issues were added to my lists below.

I'll now move on to programmatic assessment.

In [89]:
twitter_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

Here, I can see that timestamp is an integer instead of a datetime object (quality issue). I can also see that the name, doggo, floofer, pupper, and puppo columns do not have any null values even though they have several rows listed as 'None'. Additionally, I see that in_reply_to_status_id and in_reply_to_user_id both have 78 non-null values, meaning that 78 of these tweets are replies to other tweets and need to be dropped from the analysis. Also, 181 of these entries are retweets and also need to be dropped, based on the 181 null values for retweeted_status_id and retweeted_status_user_id.

In [183]:
sum(twitter_archive.tweet_id.duplicated())

0

I ran the cell above to make sure there are no duplicated tweet ids in this table.

In [91]:
twitter_archive.rating_numerator.describe()

count    2356.000000
mean       13.126486
std        45.876648
min         0.000000
25%        10.000000
50%        11.000000
75%        12.000000
max      1776.000000
Name: rating_numerator, dtype: float64

In [126]:
twitter_archive.rating_numerator.value_counts()

12      558
11      464
10      461
13      351
9       158
8       102
7        55
14       54
5        37
6        32
3        19
4        17
1         9
2         9
420       2
0         2
15        2
75        2
80        1
20        1
24        1
26        1
44        1
50        1
60        1
165       1
84        1
88        1
144       1
182       1
143       1
666       1
960       1
1776      1
17        1
27        1
45        1
99        1
121       1
204       1
Name: rating_numerator, dtype: int64

Here, I can see that the minimum rating numerator is 0 and the max is 1776. Are these mistakes? Most of the values are between 8 and 12.

In [115]:
# query to get entry with rating numerator of 1776
twitter_archive.query('rating_numerator == 1776')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
979  749981277374128128                    NaN                  NaN   

                     timestamp  \
979  2016-07-04 15:00:45 +0000   

                                                source  \
979  <a href="https://about.twitter.com/products/tw...   

                                                  text  retweeted_status_id  \
979  This is Atticus. He's quite simply America af....                  NaN   

     retweeted_status_user_id retweeted_status_timestamp  \
979                       NaN                        NaN   

                                         expanded_urls  rating_numerator  \
979  https://twitter.com/dog_rates/status/749981277...              1776   

     rating_denominator     name doggo floofer pupper puppo  
979                  10  Atticus  None    None   None  None

In [117]:
twitter_archive.iloc[979].text

"This is Atticus. He's quite simply America af. 1776/10 https://t.co/GRXwMxLBkh"

No, this does not appear to be a mistake, so the other high values are probably accurate as well.

In [123]:
# query to get entry with rating numerator of 1776
twitter_archive.query('rating_numerator == 0')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
315   835152434251116546                    NaN                  NaN   
1016  746906459439529985           7.468859e+17         4.196984e+09   

                      timestamp  \
315   2017-02-24 15:40:31 +0000   
1016  2016-06-26 03:22:31 +0000   

                                                 source  \
315   <a href="http://twitter.com/download/iphone" r...   
1016  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
315   When you're so blinded by your systematic plag...                  NaN   
1016  PUPDATE: can't see any. Even if I could, I cou...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
315                        NaN                        NaN   
1016                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
315   https://twitter.com/dog_rates/status/835152434...                 0   
1016  https://twitter.com/dog_rates/status/746906459...                 0   

      rating_denominator  name doggo floofer pupper puppo  
315                   10  None  None    None   None  None  
1016                  10  None  None    None   None  None

In [127]:
twitter_archive.iloc[315].text

"When you're so blinded by your systematic plagiarism that you forget what day it is. 0/10 https://t.co/YbEJPkg4Ag"

The zero rating does not appear to be a mistake either (I did not look at the second one at index 1016 because it is a retweet and will be dropped).

In [106]:
twitter_archive.rating_denominator.describe()

count    2356.000000
mean       10.455433
std         6.745237
min         0.000000
25%        10.000000
50%        10.000000
75%        10.000000
max       170.000000
Name: rating_denominator, dtype: float64

In [94]:
twitter_archive.rating_denominator.value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

Clearly, most of the denominators are 10. Are the other values mistakes?

In [130]:
twitter_archive.query('rating_denominator != 10 & in_reply_to_status_id.isnull()')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
433   820690176645140481                    NaN                  NaN   
516   810984652412424192                    NaN                  NaN   
784   775096608509886464                    NaN                  NaN   
902   758467244762497024                    NaN                  NaN   
1068  740373189193256964                    NaN                  NaN   
1120  731156023742988288                    NaN                  NaN   
1165  722974582966214656                    NaN                  NaN   
1202  716439118184652801                    NaN                  NaN   
1228  713900603437621249                    NaN                  NaN   
1254  710658690886586372                    NaN                  NaN   
1274  709198395643068416                    NaN                  NaN   
1351  704054845121142784                    NaN                  NaN   
1433  697463031882764288                    NaN                  NaN   
1635  684222868335505415                    NaN                  NaN   
1662  682962037429899265                    NaN                  NaN   
1779  677716515794329600                    NaN                  NaN   
1843  675853064436391936                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   

                      timestamp  \
433   2017-01-15 17:52:40 +0000   
516   2016-12-19 23:06:23 +0000   
784   2016-09-11 22:20:06 +0000   
902   2016-07-28 01:00:57 +0000   
1068  2016-06-08 02:41:38 +0000   
1120  2016-05-13 16:15:54 +0000   
1165  2016-04-21 02:25:47 +0000   
1202  2016-04-03 01:36:11 +0000   
1228  2016-03-27 01:29:02 +0000   
1254  2016-03-18 02:46:49 +0000   
1274  2016-03-14 02:04:08 +0000   
1351  2016-02-28 21:25:30 +0000   
1433  2016-02-10 16:51:59 +0000   
1635  2016-01-05 04:00:18 +0000   
1662  2016-01-01 16:30:13 +0000   
1779  2015-12-18 05:06:23 +0000   
1843  2015-12-13 01:41:41 +0000   
2335  2015-11-16 16:11:11 +0000   

                                                 source  \
433   <a href="http://twitter.com/download/iphone" r...   
516   <a href="http://twitter.com/download/iphone" r...   
784   <a href="http://twitter.com/download/iphone" r...   
902   <a href="http://twitter.com/download/iphone" r...   
1068  <a href="http://twitter.com/download/iphone" r...   
1120  <a href="http://twitter.com/download/iphone" r...   
1165  <a href="http://twitter.com/download/iphone" r...   
1202  <a href="http://twitter.com/download/iphone" r...   
1228  <a href="http://twitter.com/download/iphone" r...   
1254  <a href="http://twitter.com/download/iphone" r...   
1274  <a href="http://twitter.com/download/iphone" r...   
1351  <a href="http://twitter.com/download/iphone" r...   
1433  <a href="http://twitter.com/download/iphone" r...   
1635  <a href="http://twitter.com/download/iphone" r...   
1662  <a href="http://twitter.com/download/iphone" r...   
1779  <a href="http://twitter.com/download/iphone" r...   
1843  <a href="http://twitter.com/download/iphone" r...   
2335  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
433   The floofs have been released I repeat the flo...                  NaN   
516   Meet Sam. She smiles 24/7 &amp; secretly aspir...                  NaN   
784   RT @dog_rates: After so many requests, this is...         7.403732e+17   
902   Why does this never happen at my front door......                  NaN   
1068  After so many requests, this is Bretagne. She ...                  NaN   
1120  Say hello to this unbelievably well behaved sq...                  NaN   
1165  Happy 4/20 from the squad! 13/10 for all https...                  NaN   
1202  This is Bluebert. He just saw that both #Final...                  NaN   
1228  Happy Saturday here's 9 puppers on a bench. 99...                  NaN   
1254  Here's a brigade of puppers. All look very pre...    

In [149]:
# get list of indices for all non-zero denominators that aren't retweets
non_ten = twitter_archive.query('rating_denominator != 10 & in_reply_to_status_id.isnull()').index.values.tolist()

# print text string for these indices
for index in non_ten:
    print (index, twitter_archive.iloc[index].text)

433 The floofs have been released I repeat the floofs have been released. 84/70 https://t.co/NIYC820tmd
516 Meet Sam. She smiles 24/7 &amp; secretly aspires to be a reindeer. 
Keep Sam smiling by clicking and sharing this link:
https://t.co/98tB8y7y7t https://t.co/LouL5vdvxx
784 RT @dog_rates: After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https:/…
902 Why does this never happen at my front door... 165/150 https://t.co/HmwrdfEfUE
1068 After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https://t.co/XAVDNDaVgQ
1120 Say hello to this unbelievably well behaved squad of doggos. 204/170 would try to pet all at once https://t.co/yGQI3He3xv
1165 Happy 4/20 from the squad! 13/10 for all https://t.co/eV1diwds8a
1202 This is Bluebert. He just saw that both #FinalFur match ups are split 50/50. Amazed af. 11/10 https://t.co/Kky1DPG4iq
1228 Happy Saturday here's 9 p

From querying above, I can see that the ratings for entries 516, 1068, 1165, 1202, 1662, 2335 are wrong. The ratings were picked up from another fraction-looking number in the text. Therefore the denominators *and* numerators for these entries will need to be corrected.


In [173]:
twitter_archive.name.value_counts()

None              745
a                  55
Charlie            12
Oliver             11
Cooper             11
Lucy               11
Penny              10
Lola               10
Tucker             10
Bo                  9
Winston             9
the                 8
Sadie               8
Daisy               7
Buddy               7
Toby                7
an                  7
Bailey              7
Jack                6
Milo                6
Dave                6
Koda                6
Stanley             6
Rusty               6
Jax                 6
Scout               6
Leo                 6
Oscar               6
Bella               6
Chester             5
                 ... 
Dallas              1
Mimosa              1
Julius              1
Emma                1
Timofy              1
Puff                1
Brooks              1
Rascal              1
Horace              1
Michelangelope      1
Bobbay              1
Lulu                1
Biden               1
Poppy               1
Pavlov    

From the value counts above, I can confirm that there are many names wrongly listed as 'None' instead of NaN, along with some names that are articles (a, an, the) instead of names and are therefore wrong as well.

In [740]:
twitter_archive.query('doggo == "doggo" & floofer == "floofer"')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
200  854010172552949760                    NaN                  NaN   

                     timestamp  \
200  2017-04-17 16:34:26 +0000   

                                                source  \
200  <a href="http://twitter.com/download/iphone" r...   

                                                  text  retweeted_status_id  \
200  At first I thought this was a shy doggo, but i...                  NaN   

     retweeted_status_user_id retweeted_status_timestamp  \
200                       NaN                        NaN   

                                         expanded_urls  rating_numerator  \
200  https://twitter.com/dog_rates/status/854010172...                11   

     rating_denominator  name  doggo  floofer pupper puppo  
200                  10  None  doggo  floofer   None  None

In [745]:
twitter_archive.iloc[200].text

"At first I thought this was a shy doggo, but it's actually a Rare Canadian Floofer Owl. Amateurs would confuse the two. 11/10 only send dogs https://t.co/TXdT3tmuYk"

In [747]:
twitter_archive.query('puppo == "puppo" & floofer == "floofer"')

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [748]:
twitter_archive.query('pupper == "pupper" & floofer == "floofer"')

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [749]:
twitter_archive.query('pupper == "pupper" & doggo == "doggo"')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
460   817777686764523521                    NaN                  NaN   
531   808106460588765185                    NaN                  NaN   
565   802265048156610565           7.331095e+17         4.196984e+09   
575   801115127852503040                    NaN                  NaN   
705   785639753186217984                    NaN                  NaN   
733   781308096455073793                    NaN                  NaN   
778   775898661951791106                    NaN                  NaN   
822   770093767776997377                    NaN                  NaN   
889   759793422261743616                    NaN                  NaN   
956   751583847268179968                    NaN                  NaN   
1063  741067306818797568                    NaN                  NaN   
1113  733109485275860992                    NaN                  NaN   

                      timestamp  \
460   2017-01-07 16:59:28 +0000   
531   2016-12-12 00:29:28 +0000   
565   2016-11-25 21:37:47 +0000   
575   2016-11-22 17:28:25 +0000   
705   2016-10-11 00:34:48 +0000   
733   2016-09-29 01:42:20 +0000   
778   2016-09-14 03:27:11 +0000   
822   2016-08-29 03:00:36 +0000   
889   2016-07-31 16:50:42 +0000   
956   2016-07-09 01:08:47 +0000   
1063  2016-06-10 00:39:48 +0000   
1113  2016-05-19 01:38:16 +0000   

                                                 source  \
460   <a href="http://twitter.com/download/iphone" r...   
531   <a href="http://twitter.com/download/iphone" r...   
565   <a href="http://twitter.com/download/iphone" r...   
575   <a href="http://twitter.com/download/iphone" r...   
705   <a href="http://twitter.com/download/iphone" r...   
733   <a href="http://vine.co" rel="nofollow">Vine -...   
778   <a href="http://twitter.com/download/iphone" r...   
822   <a href="http://twitter.com/download/iphone" r...   
889   <a href="http://twitter.com/download/iphone" r...   
956   <a href="http://twitter.com/download/iphone" r...   
1063  <a href="http://twitter.com/download/iphone" r...   
1113  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
460   This is Dido. She's playing the lead role in "...                  NaN   
531   Here we have Burke (pupper) and Dexter (doggo)...                  NaN   
565   Like doggo, like pupper version 2. Both 11/10 ...                  NaN   
575   This is Bones. He's being haunted by another d...                  NaN   
705   This is Pinot. He's a sophisticated doggo. You...                  NaN   
733   Pupper butt 1, Doggo 0. Both 12/10 https://t.c...                  NaN   
778   RT @dog_rates: Like father (doggo), like son (...         7.331095e+17   
822   RT @dog_rates: This is just downright precious...         7.410673e+17   
889   Meet Maggie &amp; Lila. Maggie is the doggo, L...                  NaN   
956   Please stop sending it pictures that don't eve...                  NaN   
1063  This is just downright precious af. 12/10 for ...                  NaN   
1113  Like father (doggo), like son (pupper). Both 1...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
460                        NaN                        NaN   
531                        NaN                        NaN   
565                        NaN                        NaN   
575                        NaN                        NaN   
705                        NaN                        NaN   
733                        NaN                        NaN   
778               4.196984e+09  2016-05-19 01:38:16 +0000   
822               4.196984e+09  2016-06-10 00:39:48 +0000   
889                        NaN                        NaN   
956                        NaN                        NaN   
1063                       NaN                        NaN   
1113                       NaN                        NaN   

                    

In [750]:
twitter_archive.query('pupper == "pupper" & puppo == "puppo"')

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

From the above cells, I see that that some tweets mention multiple things for doggo, pupper, puppo, or floofer. This seems to be mainly when there are multiple dogs in the tweet. These entries should be removed from analysis since they're associated image prediction is probably wrong (image prediction only predicts one dog type). Doggo, pupper, puppo, and floofer should then be variables in one column instead of separate columns (tidiness issue).

**`twitter_archive` quality issues:**
- some names are wrong (a, an, or the) or are listed as 'None' instead of NaN (null)
- timestamp is int64 instead of datetime object
- doggo, floofer, pupper, and puppo have many values listed as 'None' instead of NaN (null)
- 78 of the tweets are replies to other tweets
- 181 of the tweets are retweets
- entries 516, 1068, 1165, 1202, 1662, 2335 have wrong ratings (numerator and denominator)
- some of these entries don't have an image associated with them (less entries in image_predictions dataset, found below)

**`twitter_archive tidiness` issues:**
- timestamp column has day, month, and year all in same column
- doggo, floofer, pupper, and puppo have their own columns instead being of variables in one single column (but some rows have two of these variables and need to be deleted first)

I'll now move on to assessing the `image_predictions` dataframe.

In [174]:
image_predictions

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
...                  ...                                                ...   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2051  887473957103951883    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689    https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881    https://pbs.twimg.com/media/DFl99B1WsA

From my visual assessment, I see that this table has a tweet_id column that corresponds to the `twitter_archive` table; therefore the information in this `image_predictions` table should be part of the `twitter_archive` table (tidiness issue). Additionally, several columns (p1, p2, p3, p1_conf, p2_conf, p3_conf) have non-descriptive column headers and should be renamed appropriately (quality issue). I'll now assess the dataframe programmatically.

In [175]:
image_predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


There are less entries in this dataset than in the `twitter_archive` table, which means some of the tweets in  `twitter_archive` don't have images associated with them. I added this as a quality issue in the list I made earlier for the `twitter_archive` dataset.

In [182]:
sum(image_predictions.tweet_id.duplicated())

0

I ran the cell above to make sure there are no duplicated tweet ids in this table.

In [176]:
image_predictions.p1_conf.describe()

count    2075.000000
mean        0.594548
std         0.271174
min         0.044333
25%         0.364412
50%         0.588230
75%         0.843855
max         1.000000
Name: p1_conf, dtype: float64

In [177]:
image_predictions.p2_conf.describe()

count    2.075000e+03
mean     1.345886e-01
std      1.006657e-01
min      1.011300e-08
25%      5.388625e-02
50%      1.181810e-01
75%      1.955655e-01
max      4.880140e-01
Name: p2_conf, dtype: float64

In [178]:
image_predictions.p3_conf.describe()

count    2.075000e+03
mean     6.032417e-02
std      5.090593e-02
min      1.740170e-10
25%      1.622240e-02
50%      4.944380e-02
75%      9.180755e-02
max      2.734190e-01
Name: p3_conf, dtype: float64

The descriptive statistics above for p1, p2, p3 all seem to make sense - it makes sense that the values decrease as the prediction number increases.

In [179]:
image_predictions.p1.value_counts()

golden_retriever             150
Labrador_retriever           100
Pembroke                      89
Chihuahua                     83
pug                           57
chow                          44
Samoyed                       43
toy_poodle                    39
Pomeranian                    38
malamute                      30
cocker_spaniel                30
French_bulldog                26
miniature_pinscher            23
Chesapeake_Bay_retriever      23
seat_belt                     22
Staffordshire_bullterrier     20
Siberian_husky                20
German_shepherd               20
web_site                      19
Cardigan                      19
Maltese_dog                   18
beagle                        18
Eskimo_dog                    18
teddy                         18
Shetland_sheepdog             18
Lakeland_terrier              17
Rottweiler                    17
Shih-Tzu                      17
kuvasz                        16
Italian_greyhound             16
          

In [202]:
len(image_predictions.query('p1_dog == False'))

543

In [247]:
image_predictions.query('p1_dog == False').sample(random_state=42)

tweet_id                                          jpg_url  \
1975  870063196459192321  https://pbs.twimg.com/media/DBMV3NnXUAAm0Pp.jpg   

      img_num          p1   p1_conf  p1_dog        p2   p2_conf  p2_dog  \
1975        1  comic_book  0.534409   False  envelope  0.280722   False   

               p3   p3_conf  p3_dog  
1975  book_jacket  0.043786   False

From the above query in which p1_dog is False, I sampled a tweet and can see that the image is of Clifford, a cartoon dog. Because it was not a picture of a real dog, the image prediction marked it as false in this case.

In [203]:
len(image_predictions.query('p1_dog == False & p2_dog == True'))

155

In [240]:
# sample a tweet from this query
image_predictions.query('p1_dog == False & p2_dog == True').sample(random_state=42)

tweet_id                                          jpg_url  \
1042  712438159032893441  https://pbs.twimg.com/media/CeMWubMWwAA6GwF.jpg   

      img_num        p1   p1_conf  p1_dog              p2   p2_conf  p2_dog  \
1042        1  ice_bear  0.869477   False  Great_Pyrenees  0.069457    True   

                      p3  p3_conf  p3_dog  
1042  Labrador_retriever  0.02474    True

Above, I queried the table for entries in which the p1_dog was False and the p2_dog was true, and then sampled from it to see what the image was like. In this particular sample, the image does appear to be a dog, despite the 87% p1 confidence that it's a bear instead.

In [204]:
len(image_predictions.query('p1_dog == False & p2_dog == False & p3_dog == True'))

64

In [239]:
# sample a tweet from this query
image_predictions.query('p1_dog == False & p2_dog == False & p3_dog == True').sample(random_state=42)

tweet_id                                          jpg_url  \
1664  812466873996607488  https://pbs.twimg.com/media/C0Z2T_GWgAAxbL9.jpg   

      img_num          p1   p1_conf  p1_dog      p2   p2_conf  p2_dog  \
1664        1  bath_towel  0.099804   False  pillow  0.092318   False   

              p3   p3_conf  p3_dog  
1664  Great_Dane  0.078205    True

Above, I queried the table for entries in which the p1_dog and p2_dog are False and the p2_dog is true, and then sampled from it to see what the image was like. In this particular sample, the image is a dog, but clearly there are distractions in the image that lead to wrong predictions.

In [248]:
image_predictions.query('p1_conf < 0.5').sample(random_state=42)

tweet_id                                          jpg_url  \
1404  769695466921623552  https://pbs.twimg.com/media/Cq6B8V6XYAA1T1R.jpg   

      img_num   p1   p1_conf  p1_dog      p2   p2_conf  p2_dog      p3  \
1404        1  pug  0.407117    True  muzzle  0.165638   False  kuvasz   

       p3_conf  p3_dog  
1404  0.045837    True

In the sample from the above query, we can see that the p1 prediction is wrong and the p1_conf is low (less than 0.5) most likely because there is a horse in the picture that dominates the image, thereby making the prediction harder to make.

In [242]:
len(image_predictions.query('p1_dog == True & p1_conf >= 0.5'))

973

Based on my queries and random samplings above, it's clear that the image predictions are not always correct, most likely due to distractions in the image or pictures of unreal dogs. For my analysis, I will only analyze entries in which the p1_ dog is True and the p1_conf is at least 0.5 in order to minimize the number of wrongful predictions.

**`image_predictions` quality issues:**
- non-descriptive column headers (p1, p2, p3, p1_conf, p2_conf, p3_conf)
- many false predictions and low confidence levels

**`image_predictions` tidiness issues:**
- `image_predictions` table should be part of `twitter_archive` table


I'll now move on to assessing my final dataset, the `additional_twitter_info` table.

In [184]:
additional_twitter_info

contributors  coordinates          created_at display_text_range  \
0              NaN          NaN 2017-08-01 16:23:56            [0, 85]   
1              NaN          NaN 2017-08-01 00:17:27           [0, 138]   
2              NaN          NaN 2017-07-31 00:18:03           [0, 121]   
3              NaN          NaN 2017-07-30 15:58:51            [0, 79]   
4              NaN          NaN 2017-07-29 16:00:24           [0, 138]   
5              NaN          NaN 2017-07-29 00:08:17           [0, 138]   
6              NaN          NaN 2017-07-28 16:27:12           [0, 140]   
7              NaN          NaN 2017-07-28 00:22:40           [0, 118]   
8              NaN          NaN 2017-07-27 16:25:51           [0, 122]   
9              NaN          NaN 2017-07-26 15:59:51           [0, 133]   
10             NaN          NaN 2017-07-26 00:31:25           [0, 130]   
11             NaN          NaN 2017-07-25 16:11:53           [0, 107]   
12             NaN          NaN 2017-07-25 01:55:32           [0, 106]   
13             NaN          NaN 2017-07-25 00:10:02            [0, 91]   
14             NaN          NaN 2017-07-24 17:02:04           [0, 118]   
15             NaN          NaN 2017-07-24 00:19:32           [0, 138]   
16             NaN          NaN 2017-07-23 00:22:39            [0, 86]   
17             NaN          NaN 2017-07-22 16:56:37           [0, 128]   
18             NaN          NaN 2017-07-22 00:23:06            [0, 87]   
19             NaN          NaN 2017-07-20 16:49:33           [0, 127]   
20             NaN          NaN 2017-07-19 16:06:48           [0, 127]   
21             NaN          NaN 2017-07-19 03:39:09           [0, 108]   
22             NaN          NaN 2017-07-19 00:47:34            [0, 99]   
23             NaN          NaN 2017-07-18 16:08:03            [0, 88]   
24             NaN          NaN 2017-07-18 00:07:08           [0, 129]   
25             NaN          NaN 2017-07-17 16:17:36           [0, 101]   
26             NaN          NaN 2017-07-16 23:58:41           [0, 121]   
27             NaN          NaN 2017-07-16 20:14:00            [0, 71]   
28             NaN          NaN 2017-07-15 23:25:31           [0, 131]   
29             NaN          NaN 2017-07-15 16:51:35          [27, 105]   
...            ...          ...                 ...                ...   
2315           NaN          NaN 2015-11-17 00:24:19           [0, 140]   
2316           NaN          NaN 2015-11-17 00:06:54           [0, 139]   
2317           NaN          NaN 2015-11-16 23:23:41           [0, 137]   
2318           NaN          NaN 2015-11-16 21:54:18            [0, 81]   
2319           NaN          NaN 2015-11-16 21:10:36           [0, 140]   
2320           NaN          NaN 2015-11-16 20:32:58           [0, 135]   
2321           NaN          NaN 2015-11-16 20:01:42           [0, 112]   
2322           NaN          NaN 2015-11-16 19:31:45           [0, 139]   
2323           NaN          NaN 2015-11-16 16:37:02           [0, 138]   
2324           NaN          NaN 2015-11-16 16:11:11           [0, 136]   
2325           NaN          NaN 2015-11-16 15:14:19            [0, 46]   
2326           NaN          NaN 2015-11-16 14:57:41            [0, 82]   
2327           NaN          NaN 2015-11-16 04:02:55           [0, 134]   
2328           NaN          NaN 2015-11-16 03:55:04           [0, 128]   
2329           NaN          NaN 2015-11-16 03:44:34           [0, 140]   
2330           NaN          NaN 2015-11-16 03:22:39           [0, 132]   
2331           NaN          NaN 2015-11-16 02:38:37           [0, 125]   
2332           NaN          NaN 2015-11-16 01:59:36           [0, 137]   
2333           NaN          NaN 2015-11-16 01:52:02           [0, 137]   
2334           NaN          NaN 2015-11-16 01:22:45           [0, 107]   
2335           NaN          NaN 2015-11-16 01:01:59           [0, 135]   
2336           NaN          NaN 2015-11-16 00:55:59           [0, 124]   
2337      

From initial visual assessment, I can see that there are many columns in this table that are also in the twitter_archive table - these duplicated columns should be deleted from this table (tidiness issue). From programattic assessment below, I'll be able to know exactly how many columns are duplicated.

In [185]:
additional_twitter_info.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2345 entries, 0 to 2344
Data columns (total 31 columns):
contributors                     0 non-null float64
coordinates                      0 non-null float64
created_at                       2345 non-null datetime64[ns]
display_text_range               2345 non-null object
entities                         2345 non-null object
extended_entities                2069 non-null object
favorite_count                   2345 non-null int64
favorited                        2345 non-null bool
full_text                        2345 non-null object
geo                              0 non-null float64
id                               2345 non-null int64
id_str                           2345 non-null int64
in_reply_to_screen_name          78 non-null object
in_reply_to_status_id            78 non-null float64
in_reply_to_status_id_str        78 non-null float64
in_reply_to_user_id              78 non-null float64
in_reply_to_user_id_str          78 n

The desired columns from this table should be merged with the `twitter_archive` table, such as 'favorite_count' and 'retweet_count' (tidiness). The int datatypes for these columns are appropriate. In this table, the tweet id column is called 'id' instead of 'tweet_id' (quality issue). This needs to be changed to 'tweet_id' in order to merge with the `twitter_archive` table. 

In [188]:
additional_twitter_info.favorite_count.describe()

count      2345.000000
mean       8077.829424
std       12153.476849
min           0.000000
25%        1405.000000
50%        3544.000000
75%        9987.000000
max      143552.000000
Name: favorite_count, dtype: float64

In [189]:
additional_twitter_info.retweet_count.describe()

count     2345.000000
mean      3030.373561
std       5040.584448
min          0.000000
25%        606.000000
50%       1415.000000
75%       3532.000000
max      77467.000000
Name: retweet_count, dtype: float64

In [197]:
additional_twitter_info.query('retweet_count == 77467')

contributors  coordinates          created_at display_text_range  \
1028           NaN          NaN 2016-06-18 18:26:18            [0, 91]   

                                               entities  \
1028  {'hashtags': [], 'symbols': [], 'user_mentions...   

                                      extended_entities  favorite_count  \
1028  {'media': [{'id': 744234667679821824, 'id_str'...          128454   

      favorited                                          full_text  geo  \
1028      False  Here's a doggo realizing you can stand in a po...  NaN   

                            ...                          \
1028                        ...                           

      possibly_sensitive_appealable  quoted_status quoted_status_id  \
1028                            0.0            NaN              NaN   

      quoted_status_id_str  retweet_count  retweeted  retweeted_status  \
1028                   NaN          77467      False               NaN   

                                                 source truncated  \
1028  <a href="http://twitter.com/download/iphone" r...     False   

                                                   user  
1028  {'id': 4196983835, 'id_str': '4196983835', 'na...  

[1 rows x 31 columns]

In [199]:
additional_twitter_info.iloc[1028].full_text

"Here's a doggo realizing you can stand in a pool. 13/10 enlightened af (vid by Tina Conrad) https://t.co/7wE9LTEXC4"

The descriptive statistics above appear fine. Below is my short list of quality and tidiness issues for this dataset.

**`additional_twitter_info` quality issues:**
- tweet id column is named 'id' instead of 'tweet_id'


**`additional_twitter_info` tidiness issues:**
- desired columns (favorite_count, retweet_count) from this table should be merged with the `twitter_archive` table

Here is the full list of the combined quality and tidiness issues from all three tables. I'll now move on to the cleaning steps.

**Quality issues:**

`twitter_archive`:
- some names are wrong (a, an, or the) or are listed as 'None' instead of NaN (null)
- timestamp is int64 instead of datetime object
- doggo, floofer, pupper, and puppo have many values listed as 'None' instead of NaN (null)
- 78 of the tweets are replies to other tweets
- 181 of the tweets are retweets
- retweet and reply columns can be deleted once reply and retweet entries are dropped
- entries 516, 1068, 1165, 1202, 1662, 2335 have wrong ratings (numerator and denominator)
- some of these entries don't have an image associated with them (less entries in image_predictions dataset, found below)

**`image_predictions` quality issues:**
- non-descriptive column headers (p1, p2, p3, p1_conf, p2_conf, p3_conf)
- many false predictions and low confidence levels

`additional_twitter_info`:
- tweet id column is named 'id' instead of 'tweet_id'



**Tidiness issues:**
- timestamp column in `twitter_archive` has day, month, and year all in same column
- doggo, floofer, pupper, and puppo in `twitter_archive` have their own columns instead being of variables in one single column
- `image_predictions` table should be part of `twitter_archive` table
- desired columns (favorite_count, retweet_count) from `additional_twitter_info` should be merged with the `twitter_archive` table

## Clean

In [1206]:
twitter_archive_clean = twitter_archive.copy()
image_predictions_clean = image_predictions.copy()
additional_twitter_info_clean = additional_twitter_info.copy()

### Keep only original tweets (no retweets or replies) and ones that have images

#### Some tweets don't have images associated with them.

##### Define

Drop tweet_id's in `twitter_archive_clean` table that are not in the `image_predictions_clean` table.

##### Code

In [1207]:
all_ids = twitter_archive_clean.tweet_id.values.tolist()
ids_with_image = image_predictions_clean.tweet_id.values.tolist()

In [1208]:
# create list of ids to drop from twitter_archive_clean
ids_to_drop = []

for id in all_ids:
    if id not in ids_with_image:
        ids_to_drop.append(id)

len(ids_to_drop)

281

In [1209]:
len(twitter_archive_clean)

2356

In [1210]:
for index, row in twitter_archive_clean.iterrows():
    if row['tweet_id'] in ids_to_drop:
        twitter_archive_clean.drop(index, inplace=True)

##### Test

In [1211]:
len(twitter_archive_clean)

2075

In [1212]:
# confirm list of tweet ids are the same for both tables using set()

x = twitter_archive_clean.tweet_id.values.tolist()
y = image_predictions_clean.tweet_id.values.tolist()

set(x) == set(y)

True

#### Some tweets are replies and some tweets are retweets.

##### Define

Delete the entries that have a value for retweeted_status_id (181 rows) and in_reply_to_status_id (78 rows).

##### Code

In [1213]:
import numpy as np

In [1214]:
# drop rows that have a finite value for retweet_status_id
retweet_indices = twitter_archive_clean[np.isfinite(twitter_archive_clean.retweeted_status_id)].index
twitter_archive_clean = twitter_archive_clean.drop(retweet_indices)

In [1215]:
# drop rows that have a finite value for in_reply_to_status_id
reply_indices = twitter_archive_clean[np.isfinite(twitter_archive_clean.in_reply_to_status_id)].index
twitter_archive_clean = twitter_archive_clean.drop(reply_indices)

In [1216]:
twitter_archive_clean.reset_index(drop=True, inplace=True)

##### Test

In [1217]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1971 entries, 0 to 1970
Data columns (total 17 columns):
tweet_id                      1971 non-null int64
in_reply_to_status_id         0 non-null float64
in_reply_to_user_id           0 non-null float64
timestamp                     1971 non-null object
source                        1971 non-null object
text                          1971 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 1971 non-null object
rating_numerator              1971 non-null int64
rating_denominator            1971 non-null int64
name                          1971 non-null object
doggo                         1971 non-null object
floofer                       1971 non-null object
pupper                        1971 non-null object
puppo                         1971 non-null object
dtypes: float64(4), int64(3), object(10)

Confirmed that retweet and reply columns now have 0 non-null values.

### Tidiness

#### Timestamp column in `twitter_archive` has day, month, and year all in same column

##### Define

Convert timestamp column to datetime datatype (quality issue) and then parse the month, and year into separate columns using datetime library.

##### Code

In [1218]:
from datetime import datetime as dt

In [1219]:
# convert timestamp column to datetime data type
twitter_archive_clean['timestamp'] = pd.to_datetime(twitter_archive_clean['timestamp'])

In [1220]:
# parse year and month from datetime into separate columns
twitter_archive_clean['year'] = twitter_archive_clean.timestamp.dt.year
twitter_archive_clean['month'] = twitter_archive_clean.timestamp.dt.month

##### Test

In [1221]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1971 entries, 0 to 1970
Data columns (total 19 columns):
tweet_id                      1971 non-null int64
in_reply_to_status_id         0 non-null float64
in_reply_to_user_id           0 non-null float64
timestamp                     1971 non-null datetime64[ns]
source                        1971 non-null object
text                          1971 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 1971 non-null object
rating_numerator              1971 non-null int64
rating_denominator            1971 non-null int64
name                          1971 non-null object
doggo                         1971 non-null object
floofer                       1971 non-null object
pupper                        1971 non-null object
puppo                         1971 non-null object
year                          19

In [1222]:
twitter_archive_clean.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

            timestamp                                             source  \
0 2017-08-01 16:23:56  <a href="http://twitter.com/download/iphone" r...   
1 2017-08-01 00:17:27  <a href="http://twitter.com/download/iphone" r...   
2 2017-07-31 00:18:03  <a href="http://twitter.com/download/iphone" r...   
3 2017-07-30 15:58:51  <a href="http://twitter.com/download/iphone" r...   
4 2017-07-29 16:00:24  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  year  month  
0                  10   Phineas  None    None   None  None  2017      8  
1                  10     Tilly  None    None   None  None  2017      8  
2                  10    Archie  None    None   None  None  2017      7  
3                  10     Darla  None    None   None  None  2017      7  
4                  10  Franklin  None    None   None  None  2017      7

#### One variable (dog stage) in four columns (doggo, floofer, pupper, and puppo) in `twitter_archive` table.

##### Define

Delete or correct rows with more than one value for dog stage and then make doggo, floofer, pupper, and puppo as variables in a column called 'dog_stage' using melt function.

##### Code

In [1225]:
twitter_archive_clean.query('doggo == "doggo" & floofer == "floofer"')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
152  854010172552949760                    NaN                  NaN   

              timestamp                                             source  \
152 2017-04-17 16:34:26  <a href="http://twitter.com/download/iphone" r...   

                                                  text  retweeted_status_id  \
152  At first I thought this was a shy doggo, but i...                  NaN   

     retweeted_status_user_id retweeted_status_timestamp  \
152                       NaN                        NaN   

                                         expanded_urls  rating_numerator  \
152  https://twitter.com/dog_rates/status/854010172...                11   

     rating_denominator  name  doggo  floofer pupper puppo  year  month  
152                  10  None  doggo  floofer   None  None  2017      4

In [1226]:
twitter_archive_clean.iloc[152].text

"At first I thought this was a shy doggo, but it's actually a Rare Canadian Floofer Owl. Amateurs would confuse the two. 11/10 only send dogs https://t.co/TXdT3tmuYk"

The entry above is actually of an owl instead of a dog, so I'll drop it from the dataset.

In [1227]:
twitter_archive_clean.drop(152, inplace=True)
twitter_archive_clean.reset_index(drop=True, inplace=True)

In [1228]:
twitter_archive_clean.query('doggo == "doggo" & puppo == "puppo"')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
146  855851453814013952                    NaN                  NaN   

              timestamp                                             source  \
146 2017-04-22 18:31:02  <a href="http://twitter.com/download/iphone" r...   

                                                  text  retweeted_status_id  \
146  Here's a puppo participating in the #ScienceMa...                  NaN   

     retweeted_status_user_id retweeted_status_timestamp  \
146                       NaN                        NaN   

                                         expanded_urls  rating_numerator  \
146  https://twitter.com/dog_rates/status/855851453...                13   

     rating_denominator  name  doggo floofer pupper  puppo  year  month  
146                  10  None  doggo    None   None  puppo  2017      4

In [1229]:
twitter_archive_clean.iloc[146].text

"Here's a puppo participating in the #ScienceMarch. Cleverly disguising her own doggo agenda. 13/10 would keep the planet habitable for https://t.co/cMhq16isel"

I'll change the entry above to 'puppo' only, instead of both 'puppo' and 'doggo' based on the fact that it's first called a puppo and the doggo is describing the agenda.

In [1230]:
twitter_archive_clean.at[146,'doggo'] = 'None'

In [1231]:
# confirm
twitter_archive_clean.iloc[146].doggo

'None'

In [1232]:
twitter_archive_clean.query('doggo == "doggo" & pupper == "pupper"')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
336  817777686764523521                    NaN                  NaN   
393  808106460588765185                    NaN                  NaN   
420  801115127852503040                    NaN                  NaN   
505  785639753186217984                    NaN                  NaN   
647  759793422261743616                    NaN                  NaN   
699  751583847268179968                    NaN                  NaN   
788  741067306818797568                    NaN                  NaN   
834  733109485275860992                    NaN                  NaN   

              timestamp                                             source  \
336 2017-01-07 16:59:28  <a href="http://twitter.com/download/iphone" r...   
393 2016-12-12 00:29:28  <a href="http://twitter.com/download/iphone" r...   
420 2016-11-22 17:28:25  <a href="http://twitter.com/download/iphone" r...   
505 2016-10-11 00:34:48  <a href="http://twitter.com/download/iphone" r...   
647 2016-07-31 16:50:42  <a href="http://twitter.com/download/iphone" r...   
699 2016-07-09 01:08:47  <a href="http://twitter.com/download/iphone" r...   
788 2016-06-10 00:39:48  <a href="http://twitter.com/download/iphone" r...   
834 2016-05-19 01:38:16  <a href="http://twitter.com/download/iphone" r...   

                                                  text  retweeted_status_id  \
336  This is Dido. She's playing the lead role in "...                  NaN   
393  Here we have Burke (pupper) and Dexter (doggo)...                  NaN   
420  This is Bones. He's being haunted by another d...                  NaN   
505  This is Pinot. He's a sophisticated doggo. You...                  NaN   
647  Meet Maggie &amp; Lila. Maggie is the doggo, L...                  NaN   
699  Please stop sending it pictures that don't eve...                  NaN   
788  This is just downright precious af. 12/10 for ...                  NaN   
834  Like father (doggo), like son (pupper). Both 1...                  NaN   

     retweeted_status_user_id retweeted_status_timestamp  \
336                       NaN                        NaN   
393                       NaN                        NaN   
420                       NaN                        NaN   
505                       NaN                        NaN   
647                       NaN                        NaN   
699                       NaN                        NaN   
788                       NaN                        NaN   
834                       NaN                        NaN   

                                         expanded_urls  rating_numerator  \
336  https://twitter.com/dog_rates/status/817777686...                13   
393  https://twitter.com/dog_rates/status/808106460...                12   
420  https://twitter.com/dog_rates/status/801115127...                12   
505  https://twitter.com/dog_rates/status/785639753...                10   
647  https://twitter.com/dog_rates/status/759793422...                12   
699  https://twitter.com/dog_rates/status/751583847...                 5   
788  https://twitter.com/dog_rates/status/741067306...                12   
834  https://twitter.com/dog_rates/status/733109485...                12   

     rating_denominator    name  doggo floofer  pupper puppo  year  month  
336                  10    Dido  doggo    None  pupper  None  2017      1  
393                  10    None  doggo    None  pupper  None  2016     12  
420                  10   Bones  doggo    None  pupper  None  2016     11  
505                  10   Pinot  doggo    None  pupper  None  2016     10  
647                  10  Maggie  doggo    None  pupper  None  2016      7  
699                  10    None  doggo    None  pupper  None  2016      7  
788                  10    just  doggo    None  pupper  None  2016      6  
834                  10    None  doggo    None  pupper  None  2016      5

In [1233]:
twitter_archive_clean.iloc[336].text

'This is Dido. She\'s playing the lead role in "Pupper Stops to Catch Snow Before Resuming Shadow Box with Dried Apple." 13/10 (IG: didodoggo) https://t.co/m7isZrOBX7'

Looks like the above entry should be set to none for all dog stages.

In [1234]:
twitter_archive_clean.at[336,'doggo'] = 'None'
twitter_archive_clean.at[336,'pupper'] = 'None'

In [1235]:
twitter_archive_clean.loc[420].text

"This is Bones. He's being haunted by another doggo of roughly the same size. 12/10 deep breaths pupper everything's fine https://t.co/55Dqe0SJNj"

For this entry, it appears that Bones is referred to as a pupper and the other dog is a doggo, so I'll set doggo to 'None'.

In [1236]:
twitter_archive_clean.at[420,'doggo'] = 'None'

In [1237]:
twitter_archive_clean.loc[505].text

"This is Pinot. He's a sophisticated doggo. You can tell by the hat. Also pointier than your average pupper. Still 10/10 would pet cautiously https://t.co/f2wmLZTPHd"

The entry above should just be doggo and not pupper. I'll change the value for pupper to 'None'.

In [1238]:
twitter_archive_clean.at[505,'pupper'] = 'None'

In [1239]:
twitter_archive_clean.loc[699].text

"Please stop sending it pictures that don't even have a doggo or pupper in them. Churlish af. 5/10 neat couch tho https://t.co/u2c9c7qSg8"

This entry should be 'None' for all dog stages.

In [1240]:
twitter_archive_clean.at[699,'pupper'] = 'None'
twitter_archive_clean.at[699,'doggo'] = 'None'

In [1241]:
twitter_archive_clean.loc[788].text

'This is just downright precious af. 12/10 for both pupper and doggo https://t.co/o5J479bZUC'

The above tweet and the rest of the tweets from visual inspection of the query appear to have two dogs in them. I'll drop these before moving on (entries 393, 647, 788, 834).

In [1242]:
twitter_archive_clean.drop([393,647,788,834], inplace=True)
twitter_archive_clean.reset_index(drop=True, inplace=True)

In [1243]:
twitter_archive_clean.query('puppo == "puppo" & pupper == "pupper"')

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo, year, month]
Index: []

In [1244]:
twitter_archive_clean.query('floofer == "floofer" & pupper == "pupper"')

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo, year, month]
Index: []

In [1245]:
twitter_archive_clean.query('puppo == "puppo" & floofer == "floofer"')

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo, year, month]
Index: []

Now all entries should be associated with only one of the dog stages, or none at all (i.e. doggo, floofer, pupper, or puppo).

In [1246]:
# get list of all column_names
column_names = twitter_archive_clean.columns.values.tolist()

# define value variables
value_vars = ['doggo', 'floofer', 'pupper', 'puppo']

# remove value variables from column name list to get id_vars
id_vars = column_names
for x in value_vars:
    id_vars.remove(x)

# melt value variables into one column
melted = twitter_archive_clean.reset_index().melt(id_vars = id_vars, value_vars = value_vars, var_name=['dog_stage'])

In [1247]:
melted.value.value_counts()

None       7568
pupper      202
doggo        64
puppo        23
floofer       7
Name: value, dtype: int64

In [1248]:
# Delete duplicated rows with 'None' values in value column
melted = melted.drop(melted.query('dog_stage == "puppo" & value == "None"').index)
melted = melted.drop(melted.query('dog_stage == "pupper" & value == "None"').index)
melted = melted.drop(melted.query('dog_stage == "floofer" & value == "None"').index)

In [1249]:
melted.value.value_counts()

None       1902
pupper      202
doggo        64
puppo        23
floofer       7
Name: value, dtype: int64

In [1250]:
# check number of entries with no dog stage in pre-melted dataset - this is what we should have for 'None'
len(twitter_archive_clean.query('pupper == "None" & doggo == "None" & floofer == "None" & puppo == "None"'))

1670

In [1251]:
# check number of duplicated tweet_ids
sum(melted.tweet_id.duplicated())

232

In [1252]:
# drop rows with duplicated tweet_ids, keep last
melted = melted.drop(melted[melted.tweet_id.duplicated(keep='last')].index)

In [1253]:
# number for 'None' should now match length above (1670)
melted.value.value_counts()

None       1670
pupper      202
doggo        64
puppo        23
floofer       7
Name: value, dtype: int64

In [1254]:
# confirm there are no longer duplicated tweet_ids
sum(melted.tweet_id.duplicated())

0

In [1255]:
# drop dog_stage column
melted.drop(columns='dog_stage', inplace=True)

In [1256]:
# rename value column to 'dog_stage'
melted = melted.rename(columns={'value':'dog_stage'})
melted.reset_index(drop=True, inplace=True)

In [1257]:
# set twitter_archive_clean equal to copy of melted dataset
twitter_archive_clean = melted.copy()

##### Test

In [1258]:
twitter_archive_clean.dog_stage.value_counts()

None       1670
pupper      202
doggo        64
puppo        23
floofer       7
Name: dog_stage, dtype: int64

In [1259]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1966 entries, 0 to 1965
Data columns (total 16 columns):
tweet_id                      1966 non-null int64
in_reply_to_status_id         0 non-null float64
in_reply_to_user_id           0 non-null float64
timestamp                     1966 non-null datetime64[ns]
source                        1966 non-null object
text                          1966 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 1966 non-null object
rating_numerator              1966 non-null int64
rating_denominator            1966 non-null int64
name                          1966 non-null object
year                          1966 non-null int64
month                         1966 non-null int64
dog_stage                     1966 non-null object
dtypes: datetime64[ns](1), float64(4), int64(5), object(6)
memory usage: 245.8+ KB


#### `image_predictions` table should be part of `twitter_archive` table.

##### Define

Add `image_prediction_clean` columns to `twitter_archive_clean` table by merging the two dataframes on tweet_id.

##### Code

In [1260]:
# merge together on 'tweet_id'
twitter_archive_clean = pd.merge(twitter_archive_clean, image_predictions_clean, on = 'tweet_id', how='left')

##### Test

In [1261]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1966 entries, 0 to 1965
Data columns (total 27 columns):
tweet_id                      1966 non-null int64
in_reply_to_status_id         0 non-null float64
in_reply_to_user_id           0 non-null float64
timestamp                     1966 non-null datetime64[ns]
source                        1966 non-null object
text                          1966 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 1966 non-null object
rating_numerator              1966 non-null int64
rating_denominator            1966 non-null int64
name                          1966 non-null object
year                          1966 non-null int64
month                         1966 non-null int64
dog_stage                     1966 non-null object
jpg_url                       1966 non-null object
img_num                       1966

In [1262]:
twitter_archive_clean.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

            timestamp                                             source  \
0 2017-08-01 16:23:56  <a href="http://twitter.com/download/iphone" r...   
1 2017-08-01 00:17:27  <a href="http://twitter.com/download/iphone" r...   
2 2017-07-31 00:18:03  <a href="http://twitter.com/download/iphone" r...   
3 2017-07-30 15:58:51  <a href="http://twitter.com/download/iphone" r...   
4 2017-07-29 16:00:24  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls   ...    img_num  \
0  https://twitter.com/dog_rates/status/892420643...   ...          1   
1  https://twitter.com/dog_rates/status/892177421...   ...          1   
2  https://twitter.com/dog_rates/status/891815181...   ...          1   
3  https://twitter.com/dog_rates/status/891689557...   ...          1   
4  https://twitter.com/dog_rates/status/891327558...   ...          2   

            p1   p1_conf  p1_dog                  p2   p2_conf p2_dog  \
0       orange  0.097049   False               bagel  0.085851  False   
1    Chihuahua  0.323581    True            Pekinese  0.090647   True   
2    Chihuahua  0.716012    True            malamute  0.078253   True   
3  paper_towel  0.170278   False  Labrador_retriever  0.168086   True   
4       basset  0.555712    True    English_springer  0.225770   True   

                            p3   p3_conf  p3_dog  
0                       banana  0.076110   False  
1                     papillon  0.068957    True  
2                       kelpie  0.031379    True  
3                      spatula  0.040836   False  
4  German_short-haired_pointer  0.175219    True  

[5 rows x 27 columns]

#### Desired columns (favorite_count, retweet_count) from `additional_twitter_info` should be part of the `twitter_archive` table.

##### Define

First rename 'id' column in `additional_twitter_info_clean` table to 'tweet_id' and then merge favorite_count and retweet_count columns with `twitter_archive_clean`.

##### Code

In [1263]:
# rename 'id' column to 'tweet_id'
additional_twitter_info_clean = additional_twitter_info_clean.rename(columns={'id':'tweet_id'})

In [1270]:
# select only desired 3 columns to add on
additional_info_subset = additional_twitter_info_clean.loc[:,['tweet_id', 'favorite_count', 'retweet_count']]

# merge with twitter_archive_clean
twitter_archive_clean = pd.merge(twitter_archive_clean, additional_info_subset, on='tweet_id', how='left')

##### Test

In [1271]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1966 entries, 0 to 1965
Data columns (total 29 columns):
tweet_id                      1966 non-null int64
in_reply_to_status_id         0 non-null float64
in_reply_to_user_id           0 non-null float64
timestamp                     1966 non-null datetime64[ns]
source                        1966 non-null object
text                          1966 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 1966 non-null object
rating_numerator              1966 non-null int64
rating_denominator            1966 non-null int64
name                          1966 non-null object
year                          1966 non-null int64
month                         1966 non-null int64
dog_stage                     1966 non-null object
jpg_url                       1966 non-null object
img_num                       1966

In [1272]:
twitter_archive_clean.favorite_count.describe()

count      1966.000000
mean       8893.420142
std       12627.004618
min          79.000000
25%        1930.000000
50%        4049.500000
75%       11179.000000
max      143552.000000
Name: favorite_count, dtype: float64

In [1273]:
twitter_archive_clean.retweet_count.describe()

count     1966.000000
mean      2731.735504
std       4717.107825
min         13.000000
25%        609.500000
50%       1323.500000
75%       3138.000000
max      77467.000000
Name: retweet_count, dtype: float64

### Quality

#### `twitter_archive`: wrong names in 'name' column and many listed as 'None'

##### Define

Convert names listed as 'a', 'an', 'the', or 'None' to null using str.replace

##### Code

In [257]:
def replace_name(name):
    wrong_names = ['a', 'an', 'the', 'None']
    if name in wrong_names:
        name = None

##### Test

#### `twitter_archive_clean`: dog_stage column has many values listed as 'None' instead of NaN (null)


In [1223]:
# define function to convert 'None' strings to NaN (null)
#def none_to_null(value):
   # if value == 'None':
   #     value = np.nan
   # return value

#value_vars = ['doggo', 'floofer', 'pupper', 'puppo']

#for var in value_vars:
#    twitter_archive_clean[var] = twitter_archive_clean[var].apply(none_to_null)